File: packages/snow_draw_core/lib/draw/elements/types/text/text_bounds.dart
import '../../../services/text/text_metrics_service.dart';
import '../../../types/draw_point.dart';
import '../../../types/draw_rect.dart';
import 'text_data.dart';
import 'text_layout_constants.dart';

DrawRect clampTextRectToLayout({
  required DrawRect rect,
  required DrawRect startRect,
  required DrawPoint anchor,
  required TextData data,
  TextMetricsService textMetricsService = defaultTextMetricsService,
  bool keepCenter = false,
}) {
  final resolvedTextMetricsService = resolveTextMetricsService(
    textMetricsService,
  );
  final baseLayout = resolvedTextMetricsService.measure(
    TextLayoutRequest(data: data, maxWidth: rect.width),
  );
  final horizontalPadding = resolveTextLayoutHorizontalPadding(
    baseLayout.lineHeight,
  );
  final minWidth =
      _resolveMinWidth(data, textMetricsService: resolvedTextMetricsService) +
      horizontalPadding * 2;
  final shouldClampWidth = rect.width < minWidth;

  var minX = rect.minX;
  var maxX = rect.maxX;
  if (shouldClampWidth) {
    if (keepCenter) {
      final halfWidth = minWidth / 2;
      minX = rect.centerX - halfWidth;
      maxX = rect.centerX + halfWidth;
    } else if (anchor.x <= rect.minX) {
      maxX = rect.minX + minWidth;
    } else if (anchor.x >= rect.maxX) {
      minX = rect.maxX - minWidth;
    } else {
      final ratio = _anchorRatio(anchor.x, startRect.minX, startRect.maxX);
      minX = anchor.x - minWidth * ratio;
      maxX = minX + minWidth;
    }
  }

  final layout = shouldClampWidth
      ? resolvedTextMetricsService.measure(
          TextLayoutRequest(data: data, maxWidth: minWidth),
        )
      : baseLayout;
  final minHeight = resolveTextLayoutHeight(layout);

  return DrawRect(
    minX: minX,
    minY: rect.minY,
    maxX: maxX,
    maxY: rect.minY + minHeight,
  );
}

double resolveTextLayoutHeight(TextMetrics layout) => _sanitizeExtent(
  layout.lineHeight > layout.height ? layout.lineHeight : layout.height,
);

double fitTextFontSizeToHeight({
  required TextData data,
  required double targetHeight,
  required double maxWidth,
  TextMetricsService textMetricsService = defaultTextMetricsService,
  double minFontSize = 1.0,
  int maxIterations = 8,
  double tolerance = 0.01,
}) {
  final safeWidth = _sanitizeExtent(maxWidth);
  final safeTargetHeight = _sanitizeExtent(targetHeight);
  final safeMinFontSize = _sanitizeExtent(minFontSize);
  final safeMaxIterations = maxIterations < 1 ? 1 : maxIterations;
  final safeTolerance = tolerance < 0 ? 0.0 : tolerance;
  final sanitizedFontSize = _sanitizeExtent(data.fontSize);
  final baseFontSize = sanitizedFontSize < safeMinFontSize
      ? safeMinFontSize
      : sanitizedFontSize;
  final resolvedTextMetricsService = resolveTextMetricsService(
    textMetricsService,
  );

  final baseHeight = _resolveHeight(
    data: data,
    fontSize: baseFontSize,
    maxWidth: safeWidth,
    textMetricsService: resolvedTextMetricsService,
  );
  if ((baseHeight - safeTargetHeight).abs() <= safeTolerance) {
    return baseFontSize;
  }

  final lowHeight = _resolveHeight(
    data: data,
    fontSize: safeMinFontSize,
    maxWidth: safeWidth,
    textMetricsService: resolvedTextMetricsService,
  );
  if (lowHeight >= safeTargetHeight) {
    return safeMinFontSize;
  }

  var low = safeMinFontSize;
  var high = baseFontSize < safeTargetHeight ? safeTargetHeight : baseFontSize;
  var highHeight = high == baseFontSize
      ? baseHeight
      : _resolveHeight(
          data: data,
          fontSize: high,
          maxWidth: safeWidth,
          textMetricsService: resolvedTextMetricsService,
        );

  if (highHeight < safeTargetHeight) {
    var attempts = 0;
    while (highHeight < safeTargetHeight && attempts < safeMaxIterations) {
      high *= 1.5;
      highHeight = _resolveHeight(
        data: data,
        fontSize: high,
        maxWidth: safeWidth,
        textMetricsService: resolvedTextMetricsService,
      );
      attempts += 1;
    }
    if (highHeight < safeTargetHeight) {
      return high;
    }
  }

  final span = highHeight - lowHeight;
  if (span > 0) {
    final ratio = (safeTargetHeight - lowHeight) / span;
    final estimate = low + (high - low) * ratio;
    final estimateHeight = _resolveHeight(
      data: data,
      fontSize: estimate,
      maxWidth: safeWidth,
      textMetricsService: resolvedTextMetricsService,
    );
    if ((estimateHeight - safeTargetHeight).abs() <= safeTolerance) {
      return estimate;
    }
    if (estimateHeight > safeTargetHeight) {
      high = estimate;
    } else {
      low = estimate;
    }
  }

  for (var i = 0; i < safeMaxIterations; i++) {
    final mid = (low + high) / 2;
    final height = _resolveHeight(
      data: data,
      fontSize: mid,
      maxWidth: safeWidth,
      textMetricsService: resolvedTextMetricsService,
    );
    if ((height - safeTargetHeight).abs() <= safeTolerance) {
      return mid;
    }
    if (height > safeTargetHeight) {
      high = mid;
    } else {
      low = mid;
    }
  }

  return low;
}

double _resolveMinWidth(
  TextData data, {
  required TextMetricsService textMetricsService,
}) {
  final layout = textMetricsService.measure(
    TextLayoutRequest(data: data, maxWidth: 1),
  );
  var maxLineWidth = 0.0;
  for (final line in layout.lines) {
    if (line.width > maxLineWidth) {
      maxLineWidth = line.width;
    }
  }
  if (maxLineWidth <= 0 || !maxLineWidth.isFinite) {
    return _sanitizeExtent(layout.width);
  }
  return _sanitizeExtent(maxLineWidth);
}

double _anchorRatio(double anchor, double min, double max) {
  final span = max - min;
  if (span <= 0 || !span.isFinite) {
    return 0.5;
  }
  final raw = (anchor - min) / span;
  if (!raw.isFinite) {
    return 0.5;
  }
  return raw.clamp(0.0, 1.0);
}

double _sanitizeExtent(double value) =>
    value > 0 && value.isFinite ? value : 1.0;

double _resolveHeight({
  required TextData data,
  required double fontSize,
  required double maxWidth,
  required TextMetricsService textMetricsService,
}) {
  final layout = textMetricsService.measure(
    TextLayoutRequest(
      data: data.copyWith(fontSize: fontSize),
      maxWidth: maxWidth,
    ),
  );
  return resolveTextLayoutHeight(layout);
}

File: packages/snow_draw_rust_engine/src/draw/elements/types/text/text_bounds.rs
#![allow(dead_code)]

use crate::draw::elements::types::text::text_data::{TextData, TextDataPatch};
use crate::draw::elements::types::text::text_layout_constants::resolve_text_layout_horizontal_padding;
use crate::draw::services::text::text_metrics_service::{
    TextLayoutRequest, TextMetrics, TextMetricsService, DEFAULT_TEXT_METRICS_SERVICE,
};
use crate::draw::types::draw_point::DrawPoint;
use crate::draw::types::draw_rect::DrawRect;

/// Clamps a text rectangle to the minimum width and minimum layout height.
pub fn clamp_text_rect_to_layout(
    rect: DrawRect,
    start_rect: DrawRect,
    anchor: DrawPoint,
    data: &TextData,
    text_metrics_service: Option<&dyn TextMetricsService>,
    keep_center: bool,
) -> DrawRect {
    let metrics_service = text_metrics_service.unwrap_or(&DEFAULT_TEXT_METRICS_SERVICE);

    let base_layout = metrics_service.measure(&TextLayoutRequest::new(data, rect.width()));
    let horizontal_padding = resolve_text_layout_horizontal_padding(base_layout.line_height);
    let min_width = resolve_min_width(data, metrics_service) + horizontal_padding * 2.0;
    let should_clamp_width = rect.width() < min_width;

    let mut min_x = rect.min_x;
    let mut max_x = rect.max_x;
    if should_clamp_width {
        if keep_center {
            let half_width = min_width / 2.0;
            min_x = rect.center_x() - half_width;
            max_x = rect.center_x() + half_width;
        } else if anchor.x <= rect.min_x {
            max_x = rect.min_x + min_width;
        } else if anchor.x >= rect.max_x {
            min_x = rect.max_x - min_width;
        } else {
            let ratio = anchor_ratio(anchor.x, start_rect.min_x, start_rect.max_x);
            min_x = anchor.x - min_width * ratio;
            max_x = min_x + min_width;
        }
    }

    let layout = if should_clamp_width {
        metrics_service.measure(&TextLayoutRequest::new(data, min_width))
    } else {
        base_layout
    };
    let min_height = resolve_text_layout_height(&layout);

    DrawRect::new(min_x, rect.min_y, max_x, rect.min_y + min_height)
}

/// Resolves a safe text layout height.
pub fn resolve_text_layout_height(layout: &TextMetrics) -> f64 {
    sanitize_extent(layout.line_height.max(layout.height))
}

/// Fits text font size to a target layout height for a constrained width.
pub fn fit_text_font_size_to_height(
    data: &TextData,
    target_height: f64,
    max_width: f64,
    text_metrics_service: Option<&dyn TextMetricsService>,
    min_font_size: f64,
    max_iterations: usize,
    tolerance: f64,
) -> f64 {
    let metrics_service = text_metrics_service.unwrap_or(&DEFAULT_TEXT_METRICS_SERVICE);

    let safe_width = sanitize_extent(max_width);
    let safe_target_height = sanitize_extent(target_height);
    let safe_min_font_size = sanitize_extent(min_font_size);
    let safe_max_iterations = max_iterations.max(1);
    let safe_tolerance = tolerance.max(0.0);
    let base_font_size = sanitize_extent(data.font_size).max(safe_min_font_size);

    let base_height = resolve_height(data, base_font_size, safe_width, metrics_service);
    if (base_height - safe_target_height).abs() <= safe_tolerance {
        return base_font_size;
    }

    let low_height = resolve_height(data, safe_min_font_size, safe_width, metrics_service);
    if low_height >= safe_target_height {
        return safe_min_font_size;
    }

    let mut low = safe_min_font_size;
    let mut high = if base_font_size < safe_target_height {
        safe_target_height
    } else {
        base_font_size
    };
    let mut high_height = if high == base_font_size {
        base_height
    } else {
        resolve_height(data, high, safe_width, metrics_service)
    };

    if high_height < safe_target_height {
        let mut attempts = 0usize;
        while high_height < safe_target_height && attempts < safe_max_iterations {
            high *= 1.5;
            high_height = resolve_height(data, high, safe_width, metrics_service);
            attempts += 1;
        }
        if high_height < safe_target_height {
            return high;
        }
    }

    let span = high_height - low_height;
    if span > 0.0 {
        let ratio = (safe_target_height - low_height) / span;
        let estimate = low + (high - low) * ratio;
        let estimate_height = resolve_height(data, estimate, safe_width, metrics_service);
        if (estimate_height - safe_target_height).abs() <= safe_tolerance {
            return estimate;
        }
        if estimate_height > safe_target_height {
            high = estimate;
        } else {
            low = estimate;
        }
    }

    for _ in 0..safe_max_iterations {
        let mid = (low + high) / 2.0;
        let height = resolve_height(data, mid, safe_width, metrics_service);
        if (height - safe_target_height).abs() <= safe_tolerance {
            return mid;
        }
        if height > safe_target_height {
            high = mid;
        } else {
            low = mid;
        }
    }

    low
}

fn resolve_min_width(data: &TextData, text_metrics_service: &dyn TextMetricsService) -> f64 {
    let layout = text_metrics_service.measure(&TextLayoutRequest::new(data, 1.0));
    let mut max_line_width = 0.0;
    for line in &layout.lines {
        if line.width > max_line_width {
            max_line_width = line.width;
        }
    }
    if max_line_width <= 0.0 || !max_line_width.is_finite() {
        return sanitize_extent(layout.width);
    }
    sanitize_extent(max_line_width)
}

fn anchor_ratio(anchor: f64, min: f64, max: f64) -> f64 {
    let span = max - min;
    if span <= 0.0 || !span.is_finite() {
        return 0.5;
    }
    let raw = (anchor - min) / span;
    if !raw.is_finite() {
        return 0.5;
    }
    raw.clamp(0.0, 1.0)
}

fn sanitize_extent(value: f64) -> f64 {
    if value.is_finite() && value > 0.0 {
        value
    } else {
        1.0
    }
}

fn resolve_height(
    data: &TextData,
    font_size: f64,
    max_width: f64,
    text_metrics_service: &dyn TextMetricsService,
) -> f64 {
    let adjusted_data = data.copy_with(TextDataPatch {
        font_size: Some(font_size),
        ..TextDataPatch::default()
    });
    let layout = text_metrics_service.measure(&TextLayoutRequest::new(&adjusted_data, max_width));
    resolve_text_layout_height(&layout)
}
